---
title: "Class 13"
subtitle: "Pathway Analysis from RNA-Seq data"
author: "Isabel Mejia"
format: pdf
date: "11-09-2022"
toc: true
---

## About the data 

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

>Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

# 1. Read contData and colData

```{r, message=FALSE}
library(DESeq2)
```


```{r}
colData <- read.csv("GSE37704_metadata.csv", row.names = 1)
colData
```

```{r}
countData <- read.csv("GSE37704_featurecounts.csv", row.names = 1)
head(countData)
```

Removing the troublesome data first column

```{r}
countData <- countData[,-1]
head(countData)
```

Check that row names and column names are the same. 
```{r}
all(row.names(colData) == colnames(countData))
```

All looks good apart from those zero count genes; we should remove thse 

# 2 Fix count data  

We can sum across the rows and if we get a zerio then we have no counts in any exp for a given gene
```{r}
keep.inds <- rowSums(countData) != 0
counts <- (countData[keep.inds,]) #data where 0 genes are removed

```

# QC with PCA

`prcomp()` function in base R is often used to check the data
```{r}
pca <- prcomp(t(counts), scale=TRUE)
summary(pca)


```
Our PCA score plot (aka PC1 vs PC2)

```{r}
plot(pca$x[,1], pca$x[,2], col=as.factor(colData$condition))
```


# 3. Now to run DESeq2 for differential expression analysis

```{r}
dds=DESeqDataSetFromMatrix(countData = counts,
                           colData=colData,
                           design=~condition)

dds=DESeq(dds)
res= results(dds)
head(res)
```



```{r}
summary(res)
```

```{r}
plot(res$log2FoldChange, -log(res$padj))

```

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )
mycols[res$log2FoldChange > 2] <- "blue"
mycols[res$log2FoldChange < -2] <- "blue"
mycols[res$padj > 0.05] <- "gray"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
abline(v=c(-2,+2), h=c(0.5))
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol= mapIds(org.Hs.eg.db,
                   keys=row.names(res),
                   keytype = "ENSEMBL",
                   column="SYMBOL",
                    multiVals="first")
res$entrez= mapIds(org.Hs.eg.db, 
                  keys=row.names(res),
                  keytype= "ENSEMBL",
                  column= "ENTREZID",
                  multiVals = "first")
head(res,5)

```

```{r}
library(EnhancedVolcano)

x <- as.data.frame(res)
EnhancedVolcano(x, 
                lab=x$symbol,
                x='log2FoldChange',
                y= 'pvalue')
```

```{r}
res = res[order(res$pvalue),]
write.csv(res, file="deseq_results.csv")
```

# Pathway Analysis 

We can use `gage()` with KEGG and GO

```{r, message=FALSE}
library(pathview)
library(gage)
library(gageData)
data("kegg.sets.hs")
data("sigmet.idx.hs")
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

```


What `gage()` wants as input is that vector of importance in our case that will  be the log2 fold change values. This vector should have `names()` that are entrez IDs

```{r}
foldchanges <-  res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```


```{r}
keggres = gage(foldchanges, gsets = kegg.sets.hs)

head(keggres$less, 5)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)

# Gene Ontology

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

head(gobpres$less)
```

```{r}
sig_genes <- res[res$padj <= 0.05 &  !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?

The most significant "Entities p-value" pathway is "Endosomal/Vacuolar"

Some of the most significant pathways match the previous KEGG results. 
Factors that could cause differences between the two methods would be differences in annotation methods in the two or when was the last time the database was updated




